kratos/modeler/key_plane_generation/key_plane_generation_with_refinement.cpp

```diff
@@ -0,0 +1,251 @@
+                                                const array_1d<double,3>& rGlobalVoxelSize){
+    array_1d<std::vector<double>,3> partitions;
+    array_1d<std::vector<double>,3> partitions_voxel_sizes;
+    for(int i_direction = 0 ; i_direction < 3 ; i_direction++){
+        std::vector<double>& r_i_partition = partitions[i_direction];
+        // TODO: Adding reserve
+        r_i_partition.push_back(rGlobalBoundingBox.first[i_direction]);
+        r_i_partition.push_back(rGlobalBoundingBox.second[i_direction]);
+        for(auto& area: rRefinementAreas){
+            r_i_partition.push_back(area.first[i_direction]);
+            r_i_partition.push_back(area.second[i_direction]);
+        }
+        // Now we sort the beginnig and ends and make them uniqe
+        std::sort(r_i_partition.begin(), r_i_partition.end());
+        r_i_partition.erase(std::unique(r_i_partition.begin(), r_i_partition.end()), r_i_partition.end());
+        // if there are more than 2, we loop in the inner ones and if they are too close we remove them
+    }
+    // Now we fill the proposed voxel size
+    for(int i_direction = 0 ; i_direction < 3 ; i_direction++){
+        std::vector<double>& r_i_partitions_voxel_sizes = partitions_voxel_sizes[i_direction];
+        std::vector<double>& r_i_partition = partitions[i_direction];
+        for (int i = 0;  i != (int)r_i_partition.size()-1; ++i) {
+            const double pos = 0.5*(r_i_partition[i] + r_i_partition[i+1]);
+            const double dx = ReturnLocalTheoreticalVoxelSize(rRefinementAreas, rVoxelSizes, rGlobalBoundingBox, rGlobalVoxelSize, i_direction, pos);
+            r_i_partitions_voxel_sizes.push_back(dx);
+        }
+    }
+    // Now we loop over the partitions and if 2 of them are too close we replace them for one in the middle
+    constexpr double K=0.1;
+    for(int i_direction = 0 ; i_direction < 3 ; i_direction++){
+        std::vector<double>& r_i_partitions_voxel_sizes = partitions_voxel_sizes[i_direction];
+        std::vector<double>& r_i_partition = partitions[i_direction];
+        bool keep_merging = true;
+        while(keep_merging)
+        {
+            std::vector<double> new_i_partitions;
+            std::vector<double> new_i_partitions_voxel_sizes;
+            // We add the first ones to the new partitions (leave the first point untouched)
+            new_i_partitions.push_back(*r_i_partition.begin());
+            new_i_partitions_voxel_sizes.push_back(*r_i_partitions_voxel_sizes.begin());
+            // Now we loop from the second one to the end and merge the ones in the middle
+            keep_merging = false;
+            if(r_i_partition.size()>3){
+                // The first one and the last one cannot be removed or moved.
+                for (int i = 1;  i+1 < (int) r_i_partition.size()-1; ++i) { //This is to ensure we don't remove the last one nor the first one so i+1 is the end - 1
```

Comment:
```suggestion
                for (int i = 1;  i+1 < static_cast<int>(r_i_partition.size()-1); ++i) { //This is to ensure we don't remove the last one nor the first one so i+1 is the end - 1
```

(or `std::size_t i` if not going to parallelize)

Same with all the dynamic cast in the file